File: ai/agent/reasoning.py
"""Reasoning step tracker — mirrors the frontend ReasoningTracker."""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Callable
from datetime import datetime


@dataclass
class ReasoningStep:
    type: str  # "thinking" | "tool_call" | "search" | "context_load"
    label: str
    detail: str | None = None
    timestamp: float = 0.0
    duration_ms: int | None = None


ReasoningCallback = Callable[[ReasoningStep], None]


class ReasoningTracker:
    def __init__(self):
        self.steps: list[ReasoningStep] = []
        self.callbacks: list[ReasoningCallback] = []
        self._step_start: float = 0.0

    def on_step(self, cb: ReasoningCallback) -> None:
        self.callbacks.append(cb)

    def begin(self, type_: str, label: str, detail: str | None = None) -> None:
        self._step_start = datetime.now().timestamp()
        step = ReasoningStep(type=type_, label=label, detail=detail, timestamp=self._step_start)
        self.steps.append(step)
        for cb in self.callbacks:
            cb(step)

    def end_current(self) -> None:
        if not self.steps:
            return
        last = self.steps[-1]
        last.duration_ms = int((datetime.now().timestamp() - self._step_start) * 1000)

    def get_steps(self) -> list[dict[str, Any]]:
        return [{"type": s.type, "label": s.label, "detail": s.detail, "duration_ms": s.duration_ms} for s in self.steps]

    def clear(self) -> None:
        self.steps.clear()

File: src/agent/reasoning.ts
/** Reasoning display handler (inspired by Hermes reasoning_callback). */

export interface ReasoningStep {
  type: "thinking" | "tool_call" | "search" | "context_load" | "complete"
  label: string
  detail?: string
  timestamp: number
  durationMs?: number
}

export type ReasoningCallback = (step: ReasoningStep) => void

export class ReasoningTracker {
  private steps: ReasoningStep[] = []
  private callbacks: Set<ReasoningCallback> = new Set()
  private currentStepStart = 0

  onStep(cb: ReasoningCallback): () => void {
    this.callbacks.add(cb)
    return () => this.callbacks.delete(cb)
  }

  beginThinking(label: string, detail?: string) {
    this.currentStepStart = Date.now()
    const step: ReasoningStep = {
      type: "thinking",
      label,
      detail,
      timestamp: this.currentStepStart,
    }
    this.steps.push(step)
    this.callbacks.forEach((cb) => cb(step))
  }

  beginToolCall(label: string) {
    this.currentStepStart = Date.now()
    const step: ReasoningStep = {
      type: "tool_call",
      label,
      timestamp: this.currentStepStart,
    }
    this.steps.push(step)
    this.callbacks.forEach((cb) => cb(step))
  }

  completeCurrent() {
    if (this.steps.length === 0) return
    const last = this.steps[this.steps.length - 1]
    last.durationMs = Date.now() - this.currentStepStart
    this.callbacks.forEach((cb) => cb({ ...last, type: "complete" }))
  }

  getSteps(): ReasoningStep[] {
    return [...this.steps]
  }

  clear() {
    this.steps = []
  }
}
